packages/web-media-box/src/hls-parser/tags/tagWithValueProcessors.ts

```diff
@@ -152,3 +153,38 @@ export class ExtXBitrate extends TagWithValueProcessor {
     playlist.currentBitrate = bitrate;
   }
 }
+
+export class ExtXProgramDateTime extends TagWithValueProcessor {
+  protected readonly tag = EXT_X_PROGRAM_DATE_TIME;
+
+  public process(tagValue: string, playlist: ParsedPlaylist, currentSegment: Segment): void {
+    const timestamp = Date.parse(tagValue);
+    const previousSegment = playlist.segments[playlist.segments.length - 1];
+
+    if (Number.isNaN(timestamp)) {
+      return this.warnCallback(unableToParseValueWarn(this.tag));
+    }
+
+    currentSegment.programDateTime = timestamp;
+
+    // If there are segments preceding this one without a programDateTime, we need to backfill them
+    if (previousSegment && !previousSegment.programDateTime) {
+      this.backfillProgramDateTimes(playlist, currentSegment)
+    }
+  }
+
+  private backfillProgramDateTimes(playlist: ParsedPlaylist, currentSegment: Segment): void {
+    let previousSegmentIndex = playlist.segments.indexOf(currentSegment) - 1;
+
+    // programDateTime is guranteed to be defined if we get here
+    let currentTimestamp = currentSegment.programDateTime!;
```

Comment:
I'm a TS noob, so what's the best practice here? I assume using the non-null assertion operator (`!`) is not it. `programDateTime` is guaranteed to be defined here, but TS complains because it is an optional property of `Segment`.

Reply:
`let previousSegmentIndex = playlist.segments.indexOf(currentSegment) - 1;`
    
I think this is wrong since the current segment has not been appended to the segments list yet.
We should consider the previous segment as the last segment in the list.
    
` let currentTimestamp = currentSegment.programDateTime!;`

You don't have to pass `currentSegment` as an argument to this method, you can simply pass a timestamp.

Moreover, I think you don't need a separate method for this, you can simply do it in the `process` method, with all the context you have

Reply:
yep good catch

Reply:
Instead of finding a segment (indexOf is an additional loop), use the reverse loop:


```
for (let i = playlist.segments.length - 1; i >= 0; i--) {
    //extrapolate
}

```